sensor/AdblockPlugin.js

```diff
@@ -0,0 +1,175 @@
+/*    Copyright 2016 Firewalla LLC
+ *
+ *    This program is free software: you can redistribute it and/or  modify
+ *    it under the terms of the GNU Affero General Public License, version 3,
+ *    as published by the Free Software Foundation.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+'use strict';
+
+const log = require('../net2/logger.js')(__filename);
+
+const Sensor = require('./Sensor.js').Sensor;
+
+const sem = require('../sensor/SensorEventManager.js').getInstance();
+
+const extensionManager = require('./ExtensionManager.js')
+
+const f = require('../net2/Firewalla.js');
+
+const userConfigFolder = f.getUserConfigFolder();
+const dnsmasqConfigFolder = `${userConfigFolder}/dns`;
+const deviceConfigFile = `${dnsmasqConfigFolder}/adblock_mac_set.conf`;
+const systemConfigFile = `${dnsmasqConfigFolder}/adblock_system.conf`;
+const dnsTag = "$ad_block";
+const systemLevelMac = "FF:FF:FF:FF:FF:FF";
+const updateInterval = 3600 * 1000 // once per hour
+
+const fs = require('fs');
+const Promise = require('bluebird');
+Promise.promisifyAll(fs);
+
+const DNSMASQ = require('../extension/dnsmasq/dnsmasq.js');
+const dnsmasq = new DNSMASQ();
+
+const exec = require('child-process-promise').exec;
+
+const fc = require('../net2/config.js');
+
+const spt = require('../net2/SystemPolicyTool')();
+
+class AdblockPlugin extends Sensor {
+    async run() {
+        this.systemSwitch = false;
+        this.adminSystemSwitch = false;
+        this.enabledMacAddresses = [];
+        extensionManager.registerExtension("adblock", this, {
+            applyPolicy: this.applyPolicy,
+            start: this.start,
+            stop: this.stop
+        });
+        const isPolicyEnabled = await spt.isPolicyEnabled('adblock');
```

Comment:
this should be one-time only. After this upgrade, if adblock policy:system flag is on and adblock feature flag is off, the feature flag should be kept as off